### DISCUSSION TITLE
Tokenizer - why is NumberOfTokens different from Tokens

### DISCUSSION THREAD

**Discussion by MSIH** [KeyserDSoze/Rystem.OpenAi#40] [Q&A]
when i run this, i get different values for NumberOfTokens and Tokens

```
var encoded = openAiUtility.Tokenizer
    .WithChatModel(ChatModelType.Gpt35Turbo_16K)
    .Encode(plainText);
```

![image](https://github.com/KeyserDSoze/Rystem.OpenAi/assets/9283984/27cc2397-82ff-47ed-9486-281b142abb30)

**Comment by KeyserDSoze** [comment]
Can you share the plainText value to replicate the problem?

**Comment by MSIH** [comment]
Encodedtokens = 4509
NumberOfTokens  = 4509
Tokens = count 4104

"Collective Bargaining Union Negotiations Advice | Negotiation ExpertsM4TrainingAbout our CoursesSales Negotiation TrainingProcurement Negotiation TrainingNegotiation CornerstonesNegotiation Simulation GameAdvanced Negotiation TrainingCustomized Negotiation TrainingOpen Enrollment CalendarOur TeamServicesResourcesAll ResourcesArticlesCase StudiesDefinitionsQuestionsCartoonsClientsContact Us+1 855-980-0126Tap to Email5Email+1 855-980-0126TrainingAbout our CoursesSales Negotiation TrainingProcurement Negotiation TrainingNegotiation CornerstonesNegotiation Simulation GameAdvanced Negotiation TrainingCustomized Negotiation TrainingOpen Enrollment CalendarOur TeamServicesResourcesAll ResourcesArticlesCase StudiesDefinitionsQuestionsCartoonsClientsContact UsUaUMSelect Page+1 855-980-0126Click to Email\r\nHome»Resources»Articles»Collective Bargaining Union NegotiationUpdated: 1 Sep 2021Collective Bargaining Union Negotiation\r\nSummary\r\nGain insight into best practice approaches to collective bargaining negotiation. Learn the factors that can influence the outcome of your labour or union negotiation.by Charles B. Craver\r\nCollective bargaining negotiation between labor unions and corporate employers is a specialized area in the field of general negotiations. However, the underlying legal and relationship aspects make these areas distinct. General business negotiation and lawsuit negotiations are not regulated by statutory provisions. In contrast, external laws mandate and govern a collective bargaining negotiation.Who Governs the Different Bargaining Situations?\r\nMany different statutes come into play during the negotiation process.Private sector bargaining encounters are regulated by the National Labour Relations Act (NLRA) for most workers.For railroad and airline personnel, the Railway Labour Act (RLA) regulates bargaining.The Civil Service Reform Act covers federal workers.State and local government personnel are under state public sector bargaining laws.\r\nUnder the applicable statutes, employees have the right to organize and to select exclusive bargaining agents. These agents negotiate collective agreements defining their wages, hours, and working conditions. Agents may engage in concerted activity for mutual aid and protection.\r\nFor private-sector workers, this action allows them the protected right to strike. Federal workers and most state and local employees are prohibited from striking. However, several states do permit non-essential personnel to participate in work stoppages.\r\nIndividuals who engage in a lawful economic strike may not be dismissed or otherwise disciplined for such protected activity. However, under the Mackay Radio decision of the Supreme Court, these individuals may be permanently replaced. After replacement, the strikers maintain preferential recall rights. This means employers must rehire the strikers as positions become available before hiring outside persons.Labor Unions and Negotiations\r\nLabor unions are chosen by a majority of workers in an appropriate bargaining unit. This unit may consist of homogeneous skilled workers or heterogeneous industrial workers. These workers become the bargaining agent for all of the individuals within that unit. These agents have the right to demand, on behalf of the affected employees, to bargain over factors like:WagesHoursWorking conditions \r\nOn the other hand, the NLRA specifically indicates that the duty to bargain does not require either side to agree to specific proposals or to make concessions. The participants merely have to meet at regular times and to discuss the pertinent issues in good faith.\r\nOne aspect of labor-management negotiations that is distinct involves the ongoing relationship between the sides. After completing collective discussions, the participants must continue to deal with each other:Union and management negotiators must continue to meet to resolve disagreements that may occur with respect to the application of bargaining agreement provisions.Employees and managers must work together to produce profitable goods or services if the firm is to be successful.\r\nIf union negotiators drive a hard bargain that unduly inflates labor costs, workers will be displaced by new technology. Workers may even have their jobs transferred to lower-cost areas of the U.S. or to developing countries like Mexico, China, or India. If the company treats its workers badly, morale will suffer, and good workers may seek employment elsewhere. Employees may also be less committed to firm success, causing a decrease in productivity or a reduction in work quality.The Benefits of Collective Bargaining Negotiations\r\nA factor that makes collective bargaining negotiations relatively unique from the standard negotiation definition is that it entails many issues that have to be addressed. Many types of compensation have to be discussed, including:Hourly wagesPiecework ratesFringe benefits such as pensions and health care\r\nWhat hours will the employees have to work? When will breaks and meal periods be? Almost any working condition of interest to employees might also be on the table. The expansive number of issues requires drawn-out negotiations. These talks may go on for weeks or months, as the sides try to resolve the different topics.\r\nOn the other hand, many of the bargaining subjects allow the sides to trade issues in ways that allow expanding the overall pie and maximize the joint return involved. Corporations should concede issues union leaders value more for topics management officials prefer. This permits the negotiating parties to seek win-win results that satisfy the underlying interests of both sides.The Importance of Preparation\r\nThe multifactor aspects of collective bargaining interactions make the need for thorough pre-negotiation preparation especially important. Both labor and management negotiators should sit down with the people on their respective sides before meeting with the other side. This is a chance to decide which items they should address and to ascertain their priorities. The important questions to ask in preparation include which terms are:Vital?Important?Desirable?\r\nThe teams should decide which lower value issues they are prepared to trade for preferred terms. Which topics should the team plan to raise first, and which later? Most negotiators favor starting interactions with less significant subjects. The hope is to reach tentative agreements on these topics before moving on to more important issues. This allows the team to focus initially on areas subject to joint gains while preparing to create a psychological commitment to final accords.\r\nAs the team approaches more controversial topics, those terms don’t seem as difficult as they would have had the teams begun talks with these subjects. In addition, neither side wants to see their prior tentative agreements disappear through a work stoppage. Therefore, both teams become more accommodating with respect to the controversial terms.Creative Thinking for Innovative Solutions\r\nThere will always be distributive negotiating items that both sides value. These issues generally entail monetary terms. However, even in this area, negotiators may be able to expand the pie and simultaneously enhance their respective positions.\r\nThis can be achieved if negotiators are willing to think outside the box and seek innovative solutions. For example, if profits have been decreasing, a company may offer workers a bonus instead of a pay increase. The employees get the benefit of the cash payments, but the base pay rates remain unchanged.\r\nCompanies dealing with increasing health costs could agree to larger deductibles and co-payments instead of higher employee premiums. Employee health care premiums are a difficult subject for union officials, since all workers see an immediate reduction in their take-home pay. On the other hand, workers find increased deductibles and co-payments more palatable. This is because these considerations only affect workers when they become ill. People are so relieved to have health coverage that they have less difficulty accepting the greater deductibles and co-payments.Collective Bargaining and Politics\r\nOne factor that makes collective bargaining encounters different relates to the political nature of union officials. These officials are elected leaders who generally hope for re-election. Management officials who have not taken a negotiation skills course too often forget this critical factor. As a result, these officials can embarrass their union equivalents in a public way.\r\nPolitical persons who are embarrassed before their constituents will often attempt to punish those who put them in this position. It is, therefore, important that management negotiators work to prevent such circumstances. If they have bad news for labor representatives, they should share it with them privately. They may take them outside the bargaining room or contact them on the telephone.\r\nThe union leaders understand managerial constraints and appreciate being given this information away from the eyes of unit members. Union leaders may then put on a show for their constituents at the bargaining table, but will ultimately yield to firm demands they believe to be necessary. Corporate agents should allow these leaders to continue this role during the public sessions, recognizing that it will make it easier for them to give in later.Collective Bargaining and Negotiation Anchoring\r\nAnother critical factor concerns the impact of negotiation anchoring. When one side begins with a generous initial offer, the other side is unlikely to appreciate this gift and reply in kind. Instead, one side begins to think it will do better than expected. This side then begins with a less generous opening offer.\r\nWhen I was in graduate school studying collective bargaining, I asked a friend who had been a local union president what would happen if company negotiators began with a beneficial first offer. He reacted with displeasure and suggested that such behavior would probably cause a work stoppage. Once the rank-and-file employees learned of the munificent management offer, they would raise their expectations and anticipate far better final terms.\r\nThis factor may have generated the cancelation of this year’s National Hockey League season. The team owners were clearly concerned about rising labor costs. The team owners demanded a specific division of revenues between the players and themselves. In so doing, the team owners apparently hoped to give the players no more than 54% and retain 46% for themselves. Instead of initially offering the players union 48 or 50% and allowing that side to talk them up to 53 or 54%, the owners apparently began with an offer in the 53% area.\r\nThe players and their negotiators understandably thought they might be able to get something in the 58 to 60% range. The sides reached a stalemate that they could not resolve before losing the entire season. It is, therefore, crucial for management negotiators to open with offers that are sufficiently reserved enough to leave the bargaining room once the serious talks begin. This allows the political union negotiators to talk them up and take credit for the gains they achieve.How to Avoid a Win-Loss Result\r\nNegotiating sides occasionally encounter difficult topics that neither side can surrender without a substantial loss of face. How can such issues be handled without the need for a win-loss result? If the term is not essential, the teams can resort to constructive ambiguity. This may include language pertaining to this topic that actually says nothing intelligible. Both sides are then able to claim that they did not submit. If the issue subsequently arises, the teams can try to resolve matters themselves under less difficult circumstances.\r\nIf they are unable to obtain a mutually acceptable outcome, they can invoke the contractual grievance-arbitration procedures and ask an outside neutral to decide the matter. The losing party then has someone to blame – that pointed-headed arbitrator. The labor and management representatives can then continue with their relationship without unnecessary acrimony.The History of Union Membership\r\nOver the past 50 years, the decline of union membership has greatly influenced bargaining interactions. By the mid-1950s, 35 percent of private-sector workers were union members who had their employment terms established through collective bargaining.\r\nAs the U.S. was transformed from a manufacturing to a service and white-collar economy and while American firms were directly impacted by global competition from emerging countries, the elevated labor costs associated with unionized personnel became harmful to many corporations. Non-union companies hired law firms and labor consultants to keep their firms non-union, and organized companies began to discover ways to get rid of their unions.The Future of Unions\r\nAt the same time, the Labour Board and court decisions made it easier for companies to “predict” job losses and other dire consequences associated with unionization. Unions sought to organize post-industrial entities like Wal-Mart and McDonald’s, but they used blue-collar techniques to appeal to white-collar and service personnel who thought of union membership as “lower class.” Union membership steadily lessened, resulting in a union membership rate today below 8 percent. If this trend continues and unions are unable to develop new organizing plans that appeal to post-industrial workers, they will become redundant outside such traditional industries as autos, steel, and electrical manufacturing.Opinions Around Collective Unions\r\nIn their book, What Workers Want (1999), Professors Richard Freeman and Joel Rogers discovered that over 80% of employees would like some form of collective interaction with management, with almost half of these respondents indicating an interest in traditional labor unions. On the other hand, most of the individuals indicating an interest in unionization suggested a desire for less confrontational labor-management relationships. Representative unions can no longer sit down with employer agents and simply negotiate the terms they would prefer to have.\r\nIn our global economy, they must understand the impact of their bargaining decisions on firm competitiveness. If they unduly increase labor costs or lower productivity, corporate earnings will decline, and workers will be laid off. They have to work together as partners to achieve results that reward employees for their contributions to firm success. Simultaneously, they must recognize the need to keep companies competitive.\r\nA number of successful unionized firms have taken courses together on interest-based bargaining. These courses teach labor and management representatives on how to look for ways to satisfy the underlying needs of both sides simultaneously. Despite this recent approach, its underpinnings were noted 40 years ago by Professors Richard Walton and Robert McKersie in their classic book, A Behavioural Theory of Labour Negotiations (1965). The authors discussed the need for participants to prioritize their underlying interests and seek ways to maximize the returns achieved by both sides.Collective Bargaining Decision-Making\r\nWhen complicated issues arise, teams may use separate committees to explore different options they can use to handle these matters. Companies such as The Negotiation Experts teach in their Sales Training in Seattle seminars that negotiation teams can meet away from public bargaining sessions. Groups can look for pioneering alternatives that might not have been used previously. Without the glower of public scrutiny, they can explore options that might not be ultimately adopted without the fear of embarrassment.\r\nManagement officials often complain to Labour Law teachers about how difficult it is to determine whether particular topics are mandatory bargaining subjects that must be discussed with union agents. Some subcontracting decisions that merely involve the substitution of less expensive outside workers for present employees must usually be bargained about. Other decisions involving partial department closures or other fundamental changes in the business do not have to be discussed.\r\nThe Supreme Court endeavored to draw a clear-cut line between these areas in First National Maintenance Corp. v. NLRB, 452 U.S. 666 (1981). The court indicated that “in view of an employer’s need for unencumbered decision-making, bargaining over management decisions that have a substantial impact on the continued availability of employment should be required only if the benefit, for labor-management relations and the collective bargaining process, outweighs the burden placed on the conduct of the business.”\r\nWhen basic firm decisions are based chiefly on labor cost considerations and do not entail significant changes in company operations, bargaining will generally be required. On the other hand, when the decisions do not include concern labor costs and do involve changes in basic operations, bargaining will not be necessary. Where to draw the line between required bargaining and non-mandatory bargaining is not clear. This fact should not, however, frighten management officials.The Duty to Bargain\r\nAs noted earlier, the duty to bargain does not require that either side agrees to particular demands or the making of concessions. If company leaders are considering changes that might arguably be subject to mandatory bargaining, they should resolve doubts in favor of collective negotiations.\r\nCompany leaders should advise union officials of their contemplated changes and offer to bargain. They should carefully explain the reasons for the proposed changes and ask for a union response. If the union is able to respond appropriately to their needs, company officials may decide to retain their current workers and adopt the union proposal. If union negotiators do not work to satisfy firm concerns, the company negotiators need only bargain to a good-faith impasse. At this point, they may legally effectuate their previous proposal despite union objection. They have to be sure to satisfy two crucial prerequisites to such unilateral changes. First, they must be sure they have arrived at a good-faith impasse. This is when after thorough bargaining, the sides have reached presently irreconcilable positions.\r\nWhen in doubt, they should offer to have another bargaining session to be certain they have reached this point. Second, the changes they unilaterally implement cannot be more generous to the workers than those already offered by their side at the bargaining table.Collective Bargaining Pointers\r\nPeople who must partake in collective bargaining interactions should take a good training course on negotiating if they can. It’s also helpful to read plenty of books on the negotiation process. Collective bargainers should prepare for long, drawn-out talks which will take time to develop. This is both because of the many issues to address and the political nature of union representatives.\r\nThese negotiators should also distinguish that most bargaining encounters will not be settled until shortly before the existing contract is due to expire. If labor leaders agree to terms too early, unit personnel may suspect they have become too comfortable with management and vote against contract ratification. If, on the other hand, management negotiators allow the union agents to take credit for the gains they achieved through their last-minute efforts, the affected employees are likely to be satisfied with the final results.\r\nCharles B. Craver is a Professor of Law at George Washington University.\r\n \r\n Get Newsletter!Rate this Article\r\n4.9 out of 5 from 8 responsesLoading...7 comments▲10▼ Vegeta_ble  on  7 Apr 2020 at 12:20 am \r\nVery informative! After SCOTUS’s Janus decision, things have changed. However it would seem that in some states, union membership has actually increased! From what I’ve been reading, it’s all thanks to a strong system of introducing new workers to their union representatives and explaining the benefits of union membership. Even though non-members are required to be included in any negotiations, this hasn’t resulted in members leaving to become free riders! In fact, union membership levels are even higher than before the Janus decision in some states such as Washington state! It’s worth Googling Seattle’s joint agreement with a coalition of unions to see how they are able to do so well post Janus.Reply▲1720▼ mala  on  7 Mar 2020 at 7:44 am \r\nAssume that you are on the collective bargaining panel discussing the wage talks representing the employers. What are some points you will use to defend the employers on why an increase of the wages above the inflation rate should be turned down as it is detrimental to the economy.Reply▲24▼Steam Enhancer on  7 Nov 2019 at 10:33 am \r\nI am so glad to come across this interesting and informative post. Always searching for these types of articles that contains useful stuff. Thanks for sharing.Reply▲2926▼ David Baze  on  10 Aug 2012 at 3:40 am \r\nVery informative look at collective bargainaing. As a union member i can say that we have the right to capitalize on our labor, just as the corporation has the right to capitalize. We would be a undesirable society with no middle class. Unions are still needed in my opinion, simply due to the fact that corporations are profit maximizing machines that would cut labor costs down to slavery if they could. Our econmoy is consumer driven. The more we have to spend, the more we spend. Henry Ford based his company on this premise and over paid his employees so they could afford to buy his products.Reply▲01▼ Jared replied to David Baze  on  28 Apr 2021 at 3:05 am \r\nYou have no idea what you’re blabbering about, jackass. Try having a small business that you’ve spent 18 years of your life building dismantled because the larger companies you compete with see you as a threat and are able to move the union thugs in and shoot your wages through the roof so you can no longer compete in your market. You union assholes have no idea the destruction you’re causing for small business owners. You’re a bunch of selfish and childish pricks.Reply▲2112▼ s  on  26 Dec 2011 at 6:24 pm \r\nVery informative. This article breaks down collective bargaining in a way that anyone can understand. Reply▲3612▼ Douglas Desrosiers  on  24 Apr 2008 at 6:38 pm \r\nProfessor Craver has written a remarkable article on the subject, and he must have done exhaustive research. I learned quite a few new concepts that I could not find elsewhere. The material does offer some solid advice near the end.ReplyShare your Feedback Cancel reply\r\nYour email address will not be published.Required fields are marked *\r\nComment *\r\nName *\r\nEmail *\r\nWebsite\r\n\r\n\r\nΔYOU MAY ALSO LIKEThe'FixedPie'inNegotiationsThiscasestudyshowshowalimitedfixedpiedistributivenegotiationstylecandamagenegotiationswithlabourunions.read moreHowtoBoostYourSalesSkills:7SurprisingTipsread moreJorge ChineseTeamNegotiationCross-culturalnegotiationrequiresverydetailedpreparationonculturaldifferences.ThehistoryofbusinessnegotiationshowsthatmanynegotiationsinChinafailednotread moreView More\r\nContact Us\r\nName\r\nEmail\r\nPhone\r\n\r\nMessage\r\nNewsletter Signup 3415 South Sepulveda Blvd,Suite 1100,Los Angeles 90034,USA USA:+1 855-980-0126 3rd Floor,86-90 Paul Street,London EC2A 4NE,UK UK:+44 330 027 0095 Level 25, 88 Phillip St,Sydney NSW 2000,Australia Aus:+61 (0) 2 9037 8230Refund PolicyPrivacy Policy Open in my Email App Open Gmail Copy EmailCopied!Show me a DemoSendSendWin a Free Place on a Public Course?Join Newsletter Sending  × \r\n\r\n\r\n\r\n\r\n"

**Reply by KeyserDSoze** [reply]
Sorry but I reproduced what I found on [tiktoken in python ](https://github.com/openai/tiktoken)
and I need time to study the problem, for now I create a unit test that will result in a error. I will try to solve this issue in end of august release.
